From 0b9b90a2790f073b8dfdc6ec80a931551c3500d9 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 18 Jul 2024 20:02:53 +0200
Subject: [PATCH 706/954] feat(st-clock): force clock tree for Stop2 low power
 modes

Force HSI for flexgen 2, 5 and 63 as workaround on ROM code timeout
issue for the Stop2 modes (Stop2, LP-Stop2 and LPLV-Stop2).

When PLL1 is stopped during Stop2 entry, the clock driver save the
FLEXGEN configuration and force the source to HSI, without divider,
for the clocks impacting the ROM code poling time:
* flexgen 2 : ck_icn_ddr (ck_icn_m_cpu1)
* flexgen 5 : ck_icn_nic (ck_icn_s_bootrom)
* flexgen 63 : ck_cpu1_ext2f, the CA35 bypass clock

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 00e9a111972655cab343f74e952b5f8ea2230307 feat(st-clock): force clock tree for Stop2 low power modes]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ibc98a2b0d7edf5960f97f5cf6ab0f548ffb5cc6b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/432266
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/clk/clk-stm32mp2.c | 95 ++++++++++++++++++++++++++++++++---
 1 file changed, 87 insertions(+), 8 deletions(-)

diff --git a/drivers/st/clk/clk-stm32mp2.c b/drivers/st/clk/clk-stm32mp2.c
index 23a622ad36..fcdead227d 100644
--- a/drivers/st/clk/clk-stm32mp2.c
+++ b/drivers/st/clk/clk-stm32mp2.c
@@ -24,9 +24,14 @@
 #include "clk-stm32-core.h"
 
 #ifdef IMAGE_BL31
-uint32_t saved_pll_freq1_reg;
-uint32_t saved_pll_freq2_reg;
-uint32_t saved_xbar63cfgr;
+static uint32_t saved_pll_freq1_reg;
+static uint32_t saved_pll_freq2_reg;
+static uint32_t saved_findiv2cfgr;
+static uint32_t saved_findiv5cfgr;
+static uint32_t saved_findiv63cfgr;
+static uint32_t saved_xbar2cfgr;
+static uint32_t saved_xbar5cfgr;
+static uint32_t saved_xbar63cfgr;
 #endif
 
 struct stm32_osci_dt_cfg {
@@ -2187,6 +2192,7 @@ static int wait_predivsr(uint16_t channel)
 
 	return 0;
 }
+#endif /* IMAGE_BL2 */
 
 static int wait_findivsr(uint16_t channel)
 {
@@ -2215,7 +2221,6 @@ static int wait_findivsr(uint16_t channel)
 
 	return 0;
 }
-#endif /* IMAGE_BL2 */
 
 static int wait_xbar_sts(uint16_t channel)
 {
@@ -2872,16 +2877,59 @@ int stm32mp2_pll1_disable(void)
 	saved_pll_freq1_reg = mmio_read_32(pll_freq1_reg);
 	saved_pll_freq2_reg = mmio_read_32(pll_freq2_reg);
 
-	/* save FLEXGEN63 MUX setting */
+	/* save FLEXGEN MUX setting */
+	saved_findiv2cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV2CFGR);
+	saved_findiv5cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV5CFGR);
+	saved_findiv63cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV63CFGR);
+	saved_xbar2cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_XBAR2CFGR);
+	saved_xbar5cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_XBAR5CFGR);
 	saved_xbar63cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_XBAR63CFGR);
 
+	/*
+	 * use the clock tree expected by ROM code to avoid a poling timeout
+	 * issue for wake up of Stop2 low power modes, so use HSI for
+	 * flexgen 2 : ck_icn_ddr (ck_icn_m_cpu1)
+	 * flexgen 5 : ck_icn_nic (ck_icn_s_bootrom)
+	 * flexgen 63 : ck_cpu1_ext2f, the CA35 bypass clock
+	 */
+
+	if ((saved_xbar2cfgr & RCC_XBAR2CFGR_XBAR2SEL_MASK) != XBAR_SRC_HSI) {
+		mmio_clrsetbits_32(stm32mp2_clock_data.base + RCC_XBAR2CFGR,
+				   RCC_XBAR2CFGR_XBAR2SEL_MASK, XBAR_SRC_HSI);
+		if (wait_xbar_sts(2) != 0) {
+			panic();
+		}
+		mmio_clrsetbits_32(stm32mp2_clock_data.base + RCC_FINDIV2CFGR,
+				   RCC_FINDIV2CFGR_FINDIV2_MASK, 0U);
+		if (wait_findivsr(2) != 0) {
+			panic();
+		}
+	}
+
+	if ((saved_xbar5cfgr & RCC_XBAR5CFGR_XBAR5SEL_MASK) != XBAR_SRC_HSI) {
+		mmio_clrsetbits_32(stm32mp2_clock_data.base + RCC_XBAR5CFGR,
+				   RCC_XBAR5CFGR_XBAR5SEL_MASK, XBAR_SRC_HSI);
+		if (wait_xbar_sts(5) != 0) {
+			panic();
+		}
+		mmio_clrsetbits_32(stm32mp2_clock_data.base + RCC_FINDIV5CFGR,
+				   RCC_FINDIV5CFGR_FINDIV5_MASK, 0U);
+		if (wait_findivsr(5) != 0) {
+			panic();
+		}
+	}
+
 	if ((saved_xbar63cfgr & RCC_XBAR63CFGR_XBAR63SEL_MASK) != XBAR_SRC_HSI) {
-		/* use HSI for CA35 bypass clock (FLEXGEN63 = ck_cpu1_ext2f) */
 		mmio_clrsetbits_32(stm32mp2_clock_data.base + RCC_XBAR63CFGR,
 				   RCC_XBAR63CFGR_XBAR63SEL_MASK, XBAR_SRC_HSI);
 		if (wait_xbar_sts(63) != 0) {
 			panic();
 		}
+		mmio_clrsetbits_32(stm32mp2_clock_data.base + RCC_FINDIV63CFGR,
+				   RCC_FINDIV63CFGR_FINDIV63_MASK, 0U);
+		if (wait_findivsr(63) != 0) {
+			panic();
+		}
 	}
 
 	stm32mp2_a35_ss_on_bypass();
@@ -2909,8 +2957,39 @@ int stm32mp2_pll1_enable(void)
 		panic();
 	}
 
-	/* restore FLEXGEN63 MUX setting */
-	mmio_write_32(stm32mp2_clock_data.base + RCC_XBAR63CFGR, saved_xbar63cfgr);
+	/* restore FLEXGEN MUX setting */
+	if ((saved_xbar2cfgr & RCC_XBAR2CFGR_XBAR2SEL_MASK) != XBAR_SRC_HSI) {
+		mmio_write_32(stm32mp2_clock_data.base + RCC_FINDIV2CFGR, saved_findiv2cfgr);
+		if (wait_findivsr(2) != 0) {
+			panic();
+		}
+		mmio_write_32(stm32mp2_clock_data.base + RCC_XBAR2CFGR, saved_xbar2cfgr);
+		if (wait_xbar_sts(2) != 0) {
+			panic();
+			}
+	}
+
+	if ((saved_xbar5cfgr & RCC_XBAR5CFGR_XBAR5SEL_MASK) != XBAR_SRC_HSI) {
+		mmio_write_32(stm32mp2_clock_data.base + RCC_FINDIV5CFGR, saved_findiv5cfgr);
+		if (wait_findivsr(5) != 0) {
+			panic();
+		}
+		mmio_write_32(stm32mp2_clock_data.base + RCC_XBAR5CFGR, saved_xbar5cfgr);
+		if (wait_xbar_sts(5) != 0) {
+			panic();
+		}
+	}
+
+	if ((saved_xbar63cfgr & RCC_XBAR63CFGR_XBAR63SEL_MASK) != XBAR_SRC_HSI) {
+		mmio_write_32(stm32mp2_clock_data.base + RCC_FINDIV63CFGR, saved_findiv63cfgr);
+		if (wait_findivsr(63) != 0) {
+			panic();
+		}
+		mmio_write_32(stm32mp2_clock_data.base + RCC_XBAR63CFGR, saved_xbar63cfgr);
+		if (wait_xbar_sts(63) != 0) {
+			panic();
+		}
+	}
 
 	return 0;
 #endif
-- 
2.53.0

